device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java

```diff
@@ -468,9 +470,29 @@ public void onAuthenticationSessionOpened()
     }
 
     @Override
-    public void onMessageAcknowledged(Message message)
+    public void onMessageAcknowledged(Message message, DeliveryState deliveryState)
     {
-        this.listener.onMessageSent(message, null);
+        if (deliveryState == Accepted.getInstance())
+        {
+            this.listener.onMessageSent(message, null);
+        }
+        else if (deliveryState instanceof Rejected)
+        {
+            // The message was not accepted by the server, and the reason why is found within the nested error
+            TransportException ex = AmqpsExceptionTranslator.convertFromAmqpException(((Rejected) deliveryState).getError());
+            this.listener.onMessageSent(message, ex);
+        }
+        else if (deliveryState == Released.getInstance())
+        {
+            // As per AMQP spec, this state means the message should be re-delivered to the server at a later time
+            ProtocolException protocolException = new ProtocolException("Message was released by the amqp server");
+            protocolException.setRetryable(true);
+            this.listener.onMessageSent(message, protocolException);
+        }
+        else
```

Comment:
There are at least 4 more delivery state types that are defined in the transport layer that will fall into the else statement here, can you point me to the service documentation that indicates what delivery states they may return? I am just curious.

Reply:
There isn't any documentation like that, unfortunately. For Modified/Received, I'm not sure what the SDK is supposed to do. I've never seen those delivery states returned, though

Reply:
From the official AMQP spec, though: 

• accepted: indicates successful processing at the receiver.
• rejected: indicates an invalid and unprocessable message.
• released: indicates that the message was not (and will not be) processed.
• modified: indicates that the message was modified, but not processed.
• received: indicates partial message data seen by the receiver as well as the starting point for a resumed transfer.